test/functional/plugin/lsp/watchfiles_spec.lua

```diff
@@ -0,0 +1,173 @@
+local helpers = require('test.functional.helpers')(after_each)
+
+local eq = helpers.eq
+local exec_lua = helpers.exec_lua
+local has_err = require('luassert').has.errors
+
+describe('vim.lsp._watchfiles', function()
+  before_each(helpers.clear)
+  after_each(helpers.clear)
+
+  local match = function(...)
+    return exec_lua('return require("vim.lsp._watchfiles")._match(...)', ...)
+  end
+
+  describe('glob matching', function()
+    it('should match literal strings', function()
+      eq(true, match('', ''))
+      eq(false, match('', 'a'))
+      eq(true, match('a', 'a'))
+      eq(true, match('abc', 'abc'))
+      eq(false, match('abc', 'abcdef'))
+      eq(false, match('abc', 'a'))
+      eq(false, match('a', 'b'))
+      eq(false, match('.', 'a'))
+      eq(true, match('$', '$'))
+      eq(false, match('dir/subdir', 'dir/subdir/file'))
+    end)
+
+    it('should match * wildcards', function()
+      -- eq(false, match('*', '')) -- TODO: this fails
```

Comment:
A lot of these tests are essentially documentation for how glob matching works currently since the spec isn't super clear on everything. I did find a couple of edge case like this though that I'm reasonably confident are handled incorrectly currently, but I don't think they're too critical.